File: tools/python/xen/xm/getenforce.py
#============================================================================
# This library is free software; you can redistribute it and/or
# modify it under the terms of version 2.1 of the GNU Lesser General Public
# License as published by the Free Software Foundation.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#============================================================================
# Author: Machon Gregory <mbgrego@tycho.ncsc.mil> 
#============================================================================

"""Get the current mode of the Flask XSM module.
"""

from xen.xm.opts import OptionError
from xen.xm import main as xm_main
from xen.xm.main import server
from xen.util import xsconstants

def help():
    return """
    Usage: xm getenforce

    Returns the current mode (Permissive, Enforcing) of the
    Flask XSM module."""

def getenforce():
    if xm_main.serverType == xm_main.SERVER_XEN_API:
        if xsconstants.XS_POLICY_FLASK != \
                int(server.xenapi.XSPolicy.get_xstype()):
            raise OptionError("Unsupported policy type")
        mode = int(server.xenapi.XSPolicy.getenforce())
    else:
        if server.xend.security.on() != xsconstants.XS_POLICY_FLASK:
            raise OptionError("Unsupported policy type")
        mode = server.xend.security.getenforce()
    
    if mode == 0:
        print "Permissive"
    elif mode == 1:
        print "Enforcing"

def main(argv): 
    if "-?" in argv:
        help()
        return

    if len(argv) != 1:
        raise OptionError("No arguments expected.")

    getenforce()

if __name__ == '__main__':
    try:
        main(sys.argv)
    except Exception, e:
        sys.stderr.write('Error: %s\n' % str(e))    
        sys.exit(-1)

    

File: tools/flask/utils/getenforce.c
/*
 *
 *  Author:  Machon Gregory, <mbgrego@tycho.ncsc.mil>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License version 2,
 *  as published by the Free Software Foundation.
 */

#include <stdlib.h>
#include <errno.h>
#include <stdio.h>
#include <xenctrl.h>
#include <fcntl.h>
#include <sys/mman.h>
#include <sys/stat.h>
#include <string.h>
#include <unistd.h>

static void usage (int argCnt, const char *args[])
{
    fprintf(stderr, "Usage: %s\n", args[0]);
    exit(1);
}

int main (int argCnt, const char *args[])
{
    int ret;
    xc_interface *xch = 0;

    if (argCnt != 1)
        usage(argCnt, args);

    xch = xc_interface_open(0,0,0);
    if ( !xch )
    {
        fprintf(stderr, "Unable to create interface to xenctrl: %s\n",
                strerror(errno));
        ret = -1;
        goto done;
    }

    ret = xc_flask_getenforce(xch);
    if ( ret < 0 )
    {
        errno = -ret;
        fprintf(stderr, "Unable to get enforcing mode: %s\n",
                strerror(errno));
        ret = -1;
        goto done;
    }
    else
    {
        if(ret) 
            printf("Enforcing\n");
        else
            printf("Permissive\n");
    }

done:
    if ( xch )
        xc_interface_close(xch);

    return ret;
}
